llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp

```diff
@@ -1244,17 +1247,43 @@ InstructionCost SystemZTTIImpl::getInterleavedMemoryOpCost(
   return NumVectorMemOps + NumPermutes;
 }
 
-static int getVectorIntrinsicInstrCost(Intrinsic::ID ID, Type *RetTy) {
+static int getVectorIntrinsicInstrCost(Intrinsic::ID ID, Type *RetTy, const SmallVectorImpl<Type *>& ParamTys) {
   if (RetTy->isVectorTy() && ID == Intrinsic::bswap)
     return getNumVectorRegs(RetTy); // VPERM
-  return -1;
+  else if (ID == Intrinsic::vector_reduce_add) {
+    // Retrieve number and size of elements for the vector op
+    auto *VTy = cast<FixedVectorType>(ParamTys.front());
+    unsigned numElements = VTy->getNumElements();
+    unsigned bitWidth = VTy->getScalarSizeInBits();
+    // For scalar sizes up to 128 bits, we provide a custom estimate
+    if (bitWidth <= 128) {
+      // How many elements can a single vector register hold
+      const unsigned vectorWidthInBits = 128U;
+      unsigned maxElemsPerVector = vectorWidthInBits / bitWidth;
+      // How many vector registers are needed to represent the given number of elements (V)
+      unsigned vectorRegsNeeded = getNumVectorRegs(VTy);
+      // How many instructions are needed for the final summing up of vector elements (S)
+      unsigned lastVectorHandling = std::min(numElements, 2 * (unsigned) std::round(std::log2(maxElemsPerVector)));
+      /* We need V load operations to read the input. Then, we use vector adds to
+       * create a sum vector, which takes V/2 + V/4 + V/8 + ... = V - 1 operations.
+       * Finally, we need S operations to sum up the elements of the sum vector,
+       * and a final store operation, for a total of 2 * V + S operations.
+       */
+      return 2 * vectorRegsNeeded + lastVectorHandling;
+    // For scalar sizes larger than 128 bits, we'll use the generic estimate
+    } else {
+      return -1;
+    }
+  } else {
+    return -1;
+  }
```

Comment:
Maybe just do one "return -1" at the function end..?

Reply:
sure

Reply:
done